### PR TITLE
Reconnect fix

### LINKED ISSUE #72: Normalize `reconnectTimeWait` and `maxReconnectAttempts`
Current behavior is to wait `reconnectTimeWait` for any reconnect. It should be on a per-server basis based on the last connect time. Similarly, `maxReconnectAttempts` is not per-server but an actual count of retries.

### PR DESCRIPTION
**Pull request by aricart**
FIX #72 - and normalizes `reconnectTimeWait` and `maxReconnectAttempts` to work as other clients.
Fixes the `timeout` option to account not only for the dial, but for protocol completion.

### CODE REVIEW

File: src/tcptransport.ts [hunk 1]
```diff
@@ -43,8 +45,27 @@ export class TCPTransport implements Transport {
                 this.destroy();
             }
             let connected = false;
+            let to: NodeJS.Timeout | undefined;
+            if(timeout) {
+                this.dialTime = Date.now();
```
Comment (matthiashanel) [inline src/tcptransport.ts]: Have this use of dialTime be local, such that dialTime only stores a duration?

Comment (aricart) [reply src/tcptransport.ts]: it is the initial value, that then gets used for the offset. At this point, there's no duration. When it completes we have a duration that we can work from.

File: src/protocolhandler.ts [hunk 2]
```diff
@@ -357,11 +359,16 @@ export class ProtocolHandler extends EventEmitter {
...
     private connect(): Promise<Transport> {
+        this.currentServer.lastConnect = Date.now();
         this.prepareConnection();
-        if (this.currentServer.didConnect) {
+        if (this.reconnecting) {
+            this.currentServer.reconnects += 1;
             this.client.emit('reconnecting', this.url.href);
+        } else {
+            // not on the 'reconnect' loop but honoring the reconnect policy
+            this.currentServer.connects += 1;
```
Comment (kozlovic) [inline src/protocolhandler.ts]: This is `.connects` and you comment about reconnect policy. Is this for the initial connect that follows the same logic than reconnect in term of attempts, etc..?

File: src/tcptransport.ts (hunk 1 above)
Comment (aricart) [reply src/tcptransport.ts]: done.

File: src/protocolhandler.ts (hunk 2 above)
Comment (aricart) [reply src/protocolhandler.ts]: this is left behind, I have removed them (was trying to reuse the reconnect logic), but initial connect logic either retries forever (`awaitFirstConnect`) or gives up after all servers are tried and failed.

Comment (kozlovic) [reply src/protocolhandler.ts]: Oh, I thought that the Node client would optionally use the reconnect config also for the initial connect..

Comment (aricart) [reply src/protocolhandler.ts]: no it was an in-between strat that I dropped, but didn't remove the counter.

**Pull request merged**